### DISCUSSION TITLE
How do I provide a callback for a form action?

### DISCUSSION THREAD

**Discussion by sunneydev** [vercel/next.js#61270] [Help]
### Summary

Let's say I have a `page.tsx` with the following code

```tsx
export default async function Articles() {
   async function createArticle(formData: FormData) {
       'use server'
       // create a new article based on formData
   }  
   
   return <div>
     <AddPostModal onCreate={createArticle} />
   </div>
}
```

And I have `AddPostModal` client component with the following props:
```tsx
'use client'

interface AddPostModalProps {
    onCreate: (formData: FormData) => void
}

export function AddPostModal({ onCreate }: AddPostModalProps) {
   return <form action={onCreate}>
     ....
    </form>
}
```

The following code works, but I still get an warning on `onCreate` line that props should be serializable. I'm just not sure, is this an anti pattern? should stuff be done in some other way ? or what am I missing here?

**Comment by ArnaudBand** [comment]
You can reorganise your code to avoid passing functions as props during SSR. One common method is to conditionally include or exclude certain parts of the code based on the environment (server or client).
```typescript
export default function Articles() {
  async function createArticle(formData: FormData) {
    // 'use server'
    // create a new article based on formData
  }

  return (
    <div>
      {typeof window !== 'undefined' && (
        // Render AddPostModal only on the client side
        <AddPostModal onCreate={createArticle} />
      )}
    </div>
  );
}

```
```typescript

interface AddPostModalProps {
  onCreate: (formData: FormData) => void;
}

export function AddPostModal({ onCreate }: AddPostModalProps) {
  function handleSubmit(event: React.FormEvent) {
    event.preventDefault();
    // Handle form submission or call onCreate here
  }

  return (
    <form onSubmit={handleSubmit}>
      {/* ... */}
    </form>
  );
}
```

**Reply by sunneydev** [reply]
The code I provided works fine, there are no errors, so there's no need to handle any client/server state.

My concern is about the intellisense warning. It tells me my props aren't serializable, which I understand, as `(formData: FormData) => void` is not JSON serializable, but how do I avoid this warning, while also having a callback for my form action without converting all of this to client-side code?

**Comment by hyjoong** [comment]
Are you encountering an error message like the following?
> "use server" functions are not allowed in client components. You can import them from a "use server" file instead

Server actions cannot be defined within client components. If you want to use a server action in a component where 'use client' is declared, you should import it from a separate file where only server actions using 'use server' are gathered.

```javascript
// app/actions/article.ts
export async function createArticle(formData: FormData) {
   'use server'
   // create a new article based on formData
}  
```


```javascript
'use client'

import { useTransition } from "react";
import { createArticle } from '@/actions/acticle';

export default async function Articles({formData}: FormData) {
  const [isPending, startTransition] = useTransition();

  async function handleSubmit(){
      startTransition(() => {
        createArticle(formData)
    });
  }

  return (
    <form action={handleSubmit}>
      <button disabled={isPending}>click</button>
    </form>

)

```
